### ISSUE TITLE
Issue debugging Symfony Command

### ISSUE THREAD

**Issue by mikeerickson** [felixfbecker/vscode-php-debug#352]
I am not quite sure where to turn for this, but I am trying you as you have created this plugin.  

PHP version:  
XDebug version:  

HP 7.2.7 (cli) (built: Jun 23 2018 11:20:05) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.2.0, Copyright (c) 1998-2018 Zend Technologies
    with Xdebug v2.6.1, Copyright (c) 2002-2018, by Derick Rethans
    with Zend OPcache v7.2.7, Copyright (c) 1999-2018, by Zend Technologies

Adapter version:

Latest

Your launch.json:  

```
        {
            "name": "Listen for XDebug",
            "type": "php",
            "request": "launch",
            "port": 9000
        },
```

I have also tried (If I leave out the second parameter in args, the command responds correctly in that it tells me the name (second parameter) is missing

```
       {
            "name": "Launch laravel-craftsman",
            "type": "php",
            "request": "launch",
            "program": "/Users/mikee/Documents/code/oss/laravel/laravel-craftsman/laravel-craftsman",
            "args": ["craft:class", "TestClass"],
            "runtimeArgs": [],
            "cwd": "${workspaceFolder}",
            "port": 9000
        }
```
XDebug php.ini config:

XDebug logfile (from setting `xdebug.remote_log` in php.ini):  

Adapter logfile (from setting `"log": true` in launch.json):

```bash
laravel-craftsman craft:class TestClassName
```

![image](https://user-images.githubusercontent.com/183153/56463840-e5010300-6390-11e9-9990-e84fc1ca077c.png)

**Comment by bilogic** [comment]
Hi,

I'm not familiar with ```laravel-craftsman```, but must you specify ```program``` and ```args``` in the settings? I have managed to debug all sorts of PHP code/frameworks without the need to specify them.

Further, it looks as though your ```vscode-php-debugger``` already works (i.e. able to receive debug info from your PHP server) and is actually reporting a PHP/Symfony/craftsman code logic issue. 

If I'm correct, it makes more sense to
1. Look into the stack trace to figure out which line of code was the culprit.
![image](https://user-images.githubusercontent.com/946010/56464721-44f3ba00-6422-11e9-93f9-d6ca235bc62d.png)
2. Direct the question to ```laravel-craftsman```

**Comment by mikeerickson** [comment]
I will reply w/ more Info tomorrow when I can get back to computer.  In short, Laravel craftsman is a CLI tool written in php.  The normal call in terminal is

```bash
laravel-craftsman craft:class TestService —overwrite
```

**Comment by bilogic** [comment]
I always manually start my PHP code from browser or CLI, e.g. ```php artisan db:seed```

I'm not sure what ```program``` and ```args``` are for, but it just seems very cumbersome to have to change settings for every script that you are trying to debug.

When my server's ```xdebug.so``` is enabled in ```php.ini```, any PHP script that is executed from CLI or browser will cause, ```xdebug``` to kick in where it tries to talk my ```vscode-php-debugger```. Maybe you can try ```"args": ["craft:class TestClass"]```?

**Comment by mikeerickson** [comment]
so, if you create your own artisan command, how would you setup launch.json?  It is basically the same thing.  I have tried the method of setting breakpoint in my command, leaving the program and args keys empty and the issue remains

I guess I am going to have to use PHPStorm to do debugging, all works as expected there.

**Comment by bilogic** [comment]
This is all there is in my settings
```
"configurations": [
	{
		"name": "XDebug on port 9000",
		"type": "php",
		"request": "launch",
		"port": 9000,
		"xdebugSettings": {
			"max_children": 512,
			"max_data": 512,
			"max_depth": 5
		},
		"pathMappings": {
			"<remote_path>": "<local_path>"
		}
	}
]
```

**Comment by bilogic** [comment]
Your screenshot already shows the ```xdebug``` and ```vscode-php-debug``` communicating and halting the script, halt point is denoted by the yellow arrow. The exception was not due to ```vscode-php-debug```. 

What happens when you remove ```program``` and ```args```?

**Comment by mikeerickson** [comment]
I have another command `craft:developer` which takes no arguments / options on the CLI and if I execute that command, the debugger kicks in as expected.

So, this issue likely revolves around how the args are processed?

**Comment by mikeerickson** [comment]
I am going to try your config now...

**Comment by mikeerickson** [comment]
OK, your config has the same issue as mine :-(

I have a slimmed down version as well (which is basically what you have done)

```php
        {
            "name": "Listen for XDebug",
            "type": "php",
            "request": "launch",
            "log": true,
            "port": 9000
        },
```

**Comment by bilogic** [comment]
With my config,
1. Are you starting your command manually?
2. What is the issue you are seeing? ```Runtime exception: Too many arguments```

**Comment by mikeerickson** [comment]
1. yes
2. yes

![image](https://user-images.githubusercontent.com/183153/56471888-7d889900-640c-11e9-970f-4d4b1963f8a7.png)

Additionally, I created a simple artisan command to see if it was specifically related to `laravel-craftsman` and it has the same problem.  If the command requires arguments / options, it fails

Artisan command signature:

```php
protected $signature = 'get:name {name}';
```

**Comment by mikeerickson** [comment]
Just for your etification, when debugging with PHPStorm, I setup testing a PHP Script as follows, when I start the debugging process, it properly stops at my desired breakpoint

![image](https://user-images.githubusercontent.com/183153/56471938-161f1900-640d-11e9-9334-021982cc5a2d.png)

**Comment by bilogic** [comment]
If you are seeing line 183 exception when manually starting the command, what happens when you turn off all debuggers and just start the command? Does it end with no exception?

**Comment by mikeerickson** [comment]
In case you need this, here is the information in the debugger pane for VSCode

![image](https://user-images.githubusercontent.com/183153/56471950-59798780-640d-11e9-86d9-7760c28f7187.png)

**Comment by mikeerickson** [comment]
Yes, without debugger all works as expected.  The command simply echo's whatever the user entered

![image](https://user-images.githubusercontent.com/183153/56471959-829a1800-640d-11e9-89b1-0e63767f1eb9.png)

**Comment by bilogic** [comment]
Can check through all 3 and make sure they don't have ```program``` and ```args```?
![image](https://user-images.githubusercontent.com/946010/56471965-502f0000-648b-11e9-9742-99f250dd92bd.png)

**Comment by mikeerickson** [comment]
Where is this section located?

**Comment by bilogic** [comment]
Settings, press ctrl + ,

**Comment by mikeerickson** [comment]
OK, that is what I thought you were referring to... all clean

**Comment by mikeerickson** [comment]
Are you familiar with creating Laravel commands?  not sure if you work with Laravel, thus I didnt want to assume anything :-)

**Comment by bilogic** [comment]
Yes, I write my own custom Laravel commands.

Place a breakpoint on line 75 of ```vendor/symfony/console/Input/ArgvInput.php```
The line should be ```$this->parsed = $this->tokens;``` 
Use the ```watch``` under ```vscode-php-debug``` to check if ```$this->tokens``` matches what you executed

If it matches, I have no idea what the issue is anymore.

**Comment by mikeerickson** [comment]
if you create a simple command (use attached) are you able to debug? I have removed docBlocks to simplify class.  

```php

<?php

namespace App\Console\Commands;

use Illuminate\Console\Command;

class TestOptional extends Command
{
    protected $signature = 'get:name {name} {--l|lname=erickson : User Last Name}';

    protected $description = 'Test Command';

    public function __construct()
    {
        parent::__construct();
    }

    public function handle()
    {
        $command = $this->argument('name');

        $fname = $this->option('fname');

        echo "fname: {$fname}";
    }
}


```

**Comment by mikeerickson** [comment]
Yes, it is what I expect

![image](https://user-images.githubusercontent.com/183153/56472500-ff2ff500-6413-11e9-868d-2434ebd08864.png)

**Comment by mikeerickson** [comment]
If you use the sample command above, set a breakpoint at the first line of the handle method.  In my case, it never gets there, I receive the runtime exception 

![image](https://user-images.githubusercontent.com/183153/56472513-31d9ed80-6414-11e9-8ee3-5780463a69bf.png)

**Comment by mikeerickson** [comment]
That is when I call as

```
php artisan get:name
```

If I call as

```
php artisan get:name mike
```

Then I receive 


![](https://user-images.githubusercontent.com/183153/56471888-7d889900-640c-11e9-970f-4d4b1963f8a7.png)

**Comment by bilogic** [comment]
1. Yes, I'm able to reach ```$command = $this->argument('name');``` but ```$this->option('fname');``` throws a does not exists error/exception
![image](https://user-images.githubusercontent.com/946010/56472608-0b0ecc00-6493-11e9-8870-3eae271f4eac.png)

2. What does the ```$missingArguments``` on line 75 evaluate to?

Long story short, while I can understand your logic that if ```PHPstorm``` works, so should ```vscode-php-debug```, this really does not look like anything to do with ```vscode-php-debug```'s setup etc.

I will suggest tracing through from line 75 of ```vendor/symfony/console/Input/ArgvInput.php``` til line 76 where your exception is thrown. (Strangely, why isn't it line 183 anymore?)

**Comment by bilogic** [comment]
I would also evaluate line 182's ```$all``` to have a clue.

**Comment by bilogic** [comment]
I have to go now, do update your findings, I will help again when I can.

**Comment by mikeerickson** [comment]
I am heading to family Easter plans so won’t be able to get back to this until later this evening. I am going to create a brand new empty app and try and debug. I am surprised it works for you, the command I sent is bare but doesn’t work for me

**Comment by mikeerickson** [comment]
@bilogic OK, I am back at my computer. I am going to create a brand new Laravel app command (using the example from above) and see how it goes.  I am curious what is different between our configurations (why does your work, mine not).

I doubt it is an OS related issue since my debugger kicks in, just at the error.

**Comment by mikeerickson** [comment]
@bilogic OK, I have created a clean Laravel app, created a new command (which works without debugger) and I have used your configuration in the launch.json file.

Still produces the error from above.  

So, I am really puzzled why it works on your end?  

I am going to debug the Symfony code now to see where it is getting hung up.

**Comment by mikeerickson** [comment]
@bilogic Well, after all this time, I think it *DOES* work, I just needed to continue after the initial error occurs.  Going back to the original project, I started the debug session, issued the command in the terminal, the debugger stops on the symfony error (not sure why)

So, I then clicked the | > button to continue with the debug session and then it hits my breakpoint and I am able to debug as normal.

In conclusion (and for those who may stumble upon this thread) you will likely hit the symfony error below (around line 183), but just hit the | > button to continue debugging and it should next hit your breakpoint.

![image](https://user-images.githubusercontent.com/183153/56476853-db3fd400-6452-11e9-8b7f-0e7cbbfd7891.png)

**Comment by bilogic** [comment]
Ok, I think the issue might be in your ```xdebug.ini```, here's mine.
```
[xdebug]
xdebug.remote_enable=1
xdebug.remote_host=192.168.214.3
xdebug.remote_port=9000
xdebug.remote_handler=dbgp
xdebug.remote_mode=req
xdebug.remote_autostart=0
xdebug.remote_connect_back=0
xdebug.remote_log=/var/log/xdebug.log
```

**Comment by bilogic** [comment]
Inside ```php.ini```/```xdebug.ini```, check the value of ```show_exception_trace```, it should be 0.
https://xdebug.org/docs/all_settings

**Comment by mikeerickson** [comment]
I dont have this in my `php.ini` file (I really should move this to its own `xdebug.ini` not quite sure why I don't have it this way, I do it in Docker all the time (that is a totally unrelated conversation, just thinking out loud).

So where do you have this configured?

**Comment by mikeerickson** [comment]
OK, I found it... You are indeed correct, it was not in my `php.ini` file for `xdebug` configuration
